### DISCUSSION TITLE
Questions about metrics and plot for trained models

### DISCUSSION THREAD

**Discussion by clee758** [talmolab/sleap#1526] [Help!]
Hi, I have a few questions:

1.  Is there a way to standardize the x-axis of plots to have the same scale (e.g. increase by 2 every tick), so it is easier to compare plots?
2. Is there a way to generate metrics for each node, such as the average distance, distance for the 50/75/95/99%, etc.?
3. Is there a way to export the plots as an esp file?

Thank you in advance!

**Comment by roomrys** [comment]
Hi @clee758,

1. We currently use a dynamic formula to calculate the maximum x-value to plot on the axis, so there is not an out-of-the-box way to change the x-axis of the plot at the moment.
https://github.com/talmolab/sleap/blob/ed77b49164b654bf9e43c29fcc3af73be2f8eb3b/sleap/gui/dialogs/metrics.py#L335-L336

2. The "dists" table in the analysis file has generic keys for the distances (and percentile of distances) of the instance as a whole, but we do not have any numeric return on the distance statistics for individual nodes (just the visual box plot).
https://github.com/talmolab/sleap/blob/ed77b49164b654bf9e43c29fcc3af73be2f8eb3b/sleap/nn/evals.py#L538-L551

3. Main SLEAP does not currently allow exporting the plot at all. If you have access to a plot, then you can save that plot to whichever format you would like, for example:
  ```python
  import matplotlib.pyplot as plt
  
  # Create a figure and plot some data
  fig, ax = plt.subplots()
  ax.plot([1, 2, 3], [4, 5, 6])
  
  # Save the figure as an ESP
  plt.savefig('my_plot.esp')
  ```
  
Liezl

**Reply by clee758** [reply]
Hi Liezl, thank you so much for your reply. If I'm using this code for the video plots from the Jupyter Notebook we made, what name should I put for 'my_plot' in the plt.savefig command? Should I use a string filename or a path? And if a path how should I do that?

**Reply by roomrys** [reply]
Hi @clee758,

In the case of using the Jupyter Notebook with the following `plot_distances` function:
```python
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

from pathlib import Path


def plot_distances(metrics, labels_gt, save=False):
    """Plots node distances (using matplotlib widget)."""

    fig = plt.figure()
    ax = fig.add_subplot(111)

    node_names = labels_gt.skeleton.node_names if labels_gt.skeleton else None

    dists = pd.DataFrame(metrics["dist.dists"], columns=node_names).melt(
        var_name="Part", value_name="Error"
    )

    sns.boxplot(data=dists, x="Error", y="Part", fliersize=0, ax=ax)

    sns.stripplot(
        data=dists, x="Error", y="Part", alpha=0.25, linewidth=1, jitter=0.2, ax=ax
    )

    #changing the title to the names of the videos
    title = "Node distances (ground truth vs prediction)"
    if len(labels_gt.videos) == 1:
        fn_video = Path(labels_gt.video.filename)
        title += f": {fn_video.name}"
        video_stem = fn_video.stem
    else:
        video_stem = "all_videos"

    ax.set_title(title)
    dist_1d = metrics["dist.dists"].flatten()

    xmax = np.ceil(np.ceil(np.nanpercentile(dist_1d, 95) / 5) + 1) * 5
    ax.set_xlim([0, xmax])
    ax.set_xlabel("Error (px)")

    # Get the figure filename
    fn_fig = Path(fn_models, video_stem)
    fn_fig = fn_fig.with_suffix(".png")
    fn_fig

    if save:
        plt.savefig(fn_fig)

    return fig
```

then, you could change the line that states `fn_fig = fn_fig.with_suffix(".png")` to `fn_fig = fn_fig.with_suffix(".esp")`.

Let us know how that works.

Thanks,
Liezl

**Reply by clee758** [reply]
Hi Liezl - thank you so much it worked!

I have one last question: in that plot_distances function, how can I also load a plot just for the general model in addition to models for each video? So I can also export an eps file of the general model?

**Reply by roomrys** [reply]
Hey @clee758, 

A few jargon nits: the general "model" would actually be just the general "metric". We use the same model to generate all metrics, but the video-specific metrics are only evaluating predictions/ground-truth on that video. 

Thus, the general metric (which is currently returned in main SLEAP) would be an evaluation of how the model performs on all videos in the project. An MVP script would be:
```python
from pathlib import Path
from sleap import Labels

# TODO: Change to your specific paths
fn_models = r"path/to/model/folder/that/looks/like/YYDDMM_HHMMSS.model_type.n=num_labeled"
fn_labels_gt = r"path/to/the/slp/we/want/to/evaluate"

# Create a predictor from the model
predictor = sleap.load_model(fn_models)

# Load the labels from all videos (default) and predict using model
labels_gt = sleap.load_file(fn_labels_gt)
labels_pr = predictor.predict(labels_gt)

# Calculate metrics
general_metrics = sleap.nn.evals.evaluate(labels_gt, labels_pr)

# Plot figure and save
fig = plot_distances(general_metrics, labels_gt, save=True)
fig.show()
```
which uses the `plot_distances` function defined above. I didn't explicitly test this for syntax errors, so let me know if you run into trouble!

Thanks,
Liezl

**Comment by clee758** [comment]
Hi Liezl,

Thanks for your help. I added the general_metrics variable and made a variable gfig that calls the plot_distances function like you wrote. However, I keep getting this error message: 

AttributeError                            Traceback (most recent call last)
~\AppData\Local\Temp\ipykernel_22504\728000704.py in <module>
     33     fig = plot_distances(video_metrics, labels_gt_video, save=True)
     34     fig.show()
---> 35     gfig = plot_distances(general_metrics, labels_gt, save=True)
     36     gfig.show()

~\AppData\Local\Temp\ipykernel_22504\1926365778.py in plot_distances(metrics, labels_gt, save)
     41 
     42     # Get the figure filename
---> 43     fn_fig = Path(fn_models, fn_video.stem)
     44     fn_fig = fn_fig.with_suffix(".eps")
     45     fn_fig

AttributeError: 'str' object has no attribute 'stem'

What should I do?

Thanks in advance.

**Reply by roomrys** [reply]
Sorry! My bad, for the case of passing in labels for more than one video, I had set the `fn_video` to a string instead of a `Path`. I updated the [`plot_distances` function above](https://github.com/talmolab/sleap/discussions/1526#discussioncomment-7257097) to handle this case.

**Reply by clee758** [reply]
Hi Liezl,

Thank you so much for all of your previous help!

I was just wondering how to modify the last block of code in the notebook to add an Excel sheet with metrics for the general model in addition to the individual videos?

Thank you in advance!
